feat: Refactor mech driver

mfencik opened the pull request:
After this is tested and merged, I'm aware we can do a better cleanup in workflows as we won't need undersync_device workflow anymore.
This PR needs another change in Nautobot prep_switch_interface job as from now on we can use connected_interface_uuid instead of mac_address, so please leave merging to me.

stevekeay commented on python/neutron-understack/neutron_understack/neutron_understack_mech.py:
```diff
@@ -300,45 +277,33 @@ def check_segment(self, segment):
+    def update_nautobot(self, network_id, connected_interface_uuid):
         if network_id == cfg.CONF.ml2_type_understack.provisioning_network:
-            network_name = "provisioning"
+            port_status = "Provisioning-Interface"
```
So this sets the port status to Provisioning, doesn't it have to set it back again afterwards?

mfencik replied:
this will be handled either in port delete call(not implemented yet) or it is already handled when flipping port to tenant network in prep_switch_interface job

stevekeay commented on python/neutron-understack/neutron_understack/config.py:
```diff
@@ -4,30 +4,8 @@
-        "Sleeps 5 seconds between attempts.",
-    ),
-    cfg.BoolOpt("argo_dry_run", default=True, help="Call Undersync with dry-run mode"),
-    cfg.BoolOpt("argo_force", default=False, help="Call Undersync with force mode"),
```
I added these options to the undersync API because they might be useful for troubleshooting, etc., but I don't think they need to be exposed here unless we have a specific need for them.  I can maybe see a reason to have dry_run but certainly "force" would not be a good thing to have permanently switched on.

mfencik replied:
implemented

stevekeay reviewed (commented):
Timeouts were recently added to prevent the undersync workflow from hanging indefinitely.  It might be a good idea to do that here, too.

mfencik commented on python/neutron-understack/neutron_understack/nautobot.py:
```diff
@@ -20,6 +24,39 @@ def fetch_nb_token(self):
+        args = {"url": endpoint_url, "timeout": 10}
+
+        if method != "get":
+            args["payload"] = payload
```
fixed

mfencik commented on python/neutron-understack/neutron_understack/nautobot.py:
```diff
@@ -20,6 +24,39 @@ def fetch_nb_token(self):
+    def make_api_request(self, url: str, method: str, payload: dict) -> dict:
+        endpoint_url = urljoin(self.base_url, url)
+        caller_function = inspect.stack()[self.CALLER_FRAME].function
+        api_call = getattr(self.s, method)
```
implemented

mfencik commented on python/neutron-understack/neutron_understack/nautobot.py:
```diff
@@ -20,6 +24,39 @@ def fetch_nb_token(self):
+        if method == "delete":
+            resp_data = {"status_code": resp.status_code}
+        else:
+            resp_data = resp.json()
```
implemented

skrobul commented on python/neutron-understack/neutron_understack/nautobot.py:
```diff
@@ -20,6 +24,37 @@ def fetch_nb_token(self):
+        try:
+            response.raise_for_status()
+        except HTTPError as error:
+            LOG.debug("Nautobot error: %(error)s", {"error": error})
```
```suggestion
            LOG.error("Nautobot error: %(error)s", {"error": error})
            raise from error
```

or 

```suggestion
            LOG.error("Nautobot error: %(error)s", {"error": error})
            raise NautobotError(...)
```

skrobul commented on python/neutron-understack/neutron_understack/nautobot.py:
```diff
@@ -38,52 +73,53 @@ def ucvni_create(
 
     def ucvni_delete(self, network_id):
         payload = {"id": network_id}
+        url = f"/api/plugins/undercloud-vni/ucvnis/{network_id}/"
+        resp_data = self.make_api_request(url, "delete", payload)
+        return resp_data
```
is a payload needed here?

skrobul commented on python/neutron-understack/neutron_understack/nautobot.py:
```diff
@@ -38,52 +73,53 @@ def ucvni_create(
+
+        resp_data = self.make_api_request(url, "get")
+        vlan_group_uuid = (
+            resp_data.get("relationships")
+            .get("vlan_group_to_devices")
+            .get("source")
+            .get("objects")[0]
+            .get("id")
         )
```
```suggestion
            resp_data.get("relationships", {})
            .get("vlan_group_to_devices", {})
            .get("source", {})
            .get("objects", [])[0]
            .get("id")
        )
```

skrobul commented on python/neutron-understack/neutron_understack/neutron_understack_mech.py:
```diff
@@ -214,48 +210,28 @@ def update_port_precommit(self, context):
-        )
+        network_id = context.current["network_id"]
+        connected_interface_uuid = self.fetch_connected_interface_uuid(context)
+        if not connected_interface_uuid:
```
consider adding logging

skrobul commented on python/neutron-understack/neutron_understack/neutron_understack_mech.py:
```diff
@@ -300,45 +276,33 @@ def check_segment(self, segment):
+    def fetch_connected_interface_uuid(self, context):
+        connected_interface_uuid = (
+            context.current["binding:profile"]
+            .get("local_link_information")[0]
```
```suggestion
            .get("local_link_information", [])[0]
```

skrobul commented on python/neutron-understack/neutron_understack/neutron_understack_mech.py:
```diff
@@ -300,45 +276,33 @@ def check_segment(self, segment):
+                connected_interface_uuid, port_status
+            )
+            switch_uuid = configure_port_status_data.get("device").get("id")
+            nb_vlan_group_id = self.nb.fetch_vlan_group_uuid(switch_uuid)
```
```suggestion
            return self.nb.fetch_vlan_group_uuid(switch_uuid)
```

skrobul commented on python/neutron-understack/neutron_understack/neutron_understack_mech.py:
```diff
@@ -300,45 +276,33 @@ def check_segment(self, segment):
-            service_account=cfg.CONF.ml2_type_understack.argo_workflow_sa,
-        )
-        LOG.info(f"Binding workflow submitted: {result}")
+            nb_vlan_group_id = self.nb.prep_switch_interface(
+                connected_interface_uuid, network_id
+            )
+        return nb_vlan_group_id
```
```suggestion
            return self.nb.prep_switch_interface(
                connected_interface_uuid, network_id
            )
```

skrobul commented on python/neutron-understack/neutron_understack/undersync.py:
```diff
@@ -0,0 +1,75 @@
+        }
+        return session
+
+    def undersync_post(self, action: str, uuids: str) -> requests.Response:
```
```suggestion
    def _undersync_post(self, action: str, uuids: str) -> requests.Response:
```

skrobul commented on python/neutron-understack/neutron_understack/undersync.py:
```diff
@@ -0,0 +1,75 @@
+        with file.open() as f:
+            return f.read().strip()
+
+    def log_for_status(self, response):
```
```suggestion
    def _log_for_status(self, response):
```

skrobul commented on python/neutron-understack/neutron_understack/undersync.py:
```diff
@@ -0,0 +1,80 @@
+        with file.open() as f:
+            return f.read().strip()
+
+    def _log_and_raise_for_status(self, response):
```
```suggestion
    def _log_and_raise_for_status(self, response: requests.Response):
```

skrobul reviewed (approved):
overall LGTM. Added few more suggestions for missing docstrings and type hints.

One more (optional) idea - consider updating `#update_nautobot` to return an instance of `UUID` instead of plain string, this way we add another layer of data validation which would be better than a docstring promise of returning string that happens to contain uuid.

mfencik commented:
> overall LGTM. Added few more suggestions for missing docstrings and type hints.
> 
> One more (optional) idea - consider updating `#update_nautobot` to return an instance of `UUID` instead of plain string, this way we add another layer of data validation which would be better than a docstring promise of returning string that happens to contain uuid.

that's a good idea, implemented
